orttraining/orttraining/training_ops/cuda/collective/nccl_kernels.cc

```diff
@@ -43,78 +53,201 @@ Status NcclAllGather::ComputeInternal(OpKernelContext* context) const {
+    for (int i = 0; i < context->InputCount(); i++) {
+      const Tensor* input_tensor = context->Input<Tensor>(i);
+      total_count += input_tensor->Shape().Size();
+    }
+    const int64_t alignment = size * 32;
+    const int64_t padded_count = total_count + alignment - (total_count % alignment);
+    const int64_t padded_size = padded_count * element_size;
+    auto fusion_buffer = GetScratchBuffer<void>(padded_size);
+    void* fusion_data = fusion_buffer.get();
+
+    // Calculate the range of inputs this rank will send.
+    ORT_ENFORCE(padded_count % size == 0);
+    const int64_t rank_count = padded_count / size;
+    const int64_t rank_bytes = rank_count * element_size;
+    const int64_t rank_start = rank * rank_bytes;
+    const int64_t rank_end = rank_start + rank_bytes;
+
+    // Copy this rank's inputs to fusion buffer.
+    int64_t offset = 0;
+    for (int i = 0; i < context->InputCount(); i++) {
+      const Tensor* input_tensor = context->Input<Tensor>(i);
+      const int64_t tensor_bytes = input_tensor->SizeInBytes();
+
+      // Only copy inputs this rank needs to send.
+      if (rank_start <= offset && offset < rank_end) {
+        ORT_ENFORCE(offset + tensor_bytes <= rank_end, "A single rank must be responsible for the entire tensor.");
+        void* fusion_data_at_offset = (int8_t*)fusion_data + offset;
+        const void* input_data = input_tensor->DataRaw();
+        CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(fusion_data_at_offset, input_data, tensor_bytes, cudaMemcpyDeviceToDevice));
+      }
 
-  // Calculate the range of inputs this rank will send.
-  ORT_ENFORCE(padded_count % size == 0);
-  const int64_t rank_count = padded_count / size;
-  const int64_t rank_bytes = rank_count * element_size;
-  const int64_t rank_start = rank * rank_bytes;
-  const int64_t rank_end = rank_start + rank_bytes;
+      offset += tensor_bytes;
+    }
 
-  // Copy this rank's inputs to fusion buffer.
-  int64_t offset = 0;
-  for (int i = 0; i < context->InputCount(); i++) {
-    const Tensor* input_tensor = context->Input<Tensor>(i);
-    const int64_t tensor_bytes = input_tensor->SizeInBytes();
+    // AllGather.
+    const void* fusion_data_rank_offset = (const int8_t*)fusion_data + rank_start;
+    NCCL_RETURN_IF_ERROR(ncclAllGather(fusion_data_rank_offset, fusion_data, rank_count, dtype, comm, stream));
+
+    // Copy AllGather results to outputs.
+    offset = 0;
+    for (int i = 0; i < context->InputCount(); i++) {
+      const Tensor* input_tensor = context->Input<Tensor>(i);
+      const TensorShape& input_shape = input_tensor->Shape();
+      const int64_t tensor_bytes = input_tensor->SizeInBytes();
+      Tensor* output_tensor = context->Output(i, input_shape);
+
+      // TODO: temporary hack until View is improved (it doesn't work with Alias)
+      output_tensor->SetByteOffset(input_tensor->ByteOffset());
+
+      // Only copy outputs that came from other ranks.
+      if (offset < rank_start || offset >= rank_end) {
```

Comment:
Maybe invert the condition and swap then/else branches around so that the same (rank_start <= offset && offset < rank_end) condition is repeated.